components/brave_rewards/renderer/creator_detection_agent.cc

```diff
@@ -0,0 +1,265 @@
+  return v8::ScriptCompiler::CompileFunction(context, &source, 3, parameters);
+}
+
+template <typename... Args>
+void CallFunctionWithArgs(v8::Isolate* isolate,
+                          v8::Local<v8::Function> function,
+                          Args... args) {
+  v8::Local<v8::Context> context = function->GetCreationContextChecked(isolate);
+  v8::Context::Scope context_scope(context);
+  v8::MicrotasksScope microtasks_scope(
+      context, v8::MicrotasksScope::kDoNotRunMicrotasks);
+  std::array<v8::Local<v8::Value>, sizeof...(Args)> arg_list = {
+      args.template As<v8::Value>()...};
+  std::ignore = function->Call(context, context->Global(), arg_list.size(),
+                               arg_list.empty() ? nullptr : arg_list.data());
+}
+
+}  // namespace
+
+CreatorDetectionAgent::CreatorDetectionAgent(content::RenderFrame* render_frame,
+                                             int32_t isolated_world_id)
+    : RenderFrameObserver(render_frame),
+      isolated_world_id_(isolated_world_id) {}
+
+CreatorDetectionAgent::~CreatorDetectionAgent() = default;
+
+void CreatorDetectionAgent::DidCommitProvisionalLoad(
+    ui::PageTransition transition) {
+  if (render_frame()->IsMainFrame()) {
+    MaybeInjectDetectionScript();
+  }
+}
+
+void CreatorDetectionAgent::DidFinishSameDocumentNavigation() {
+  if (render_frame()->IsMainFrame()) {
+    NotifyPageChanged();
+  }
+}
+
+void CreatorDetectionAgent::WillReleaseScriptContext(
+    v8::Local<v8::Context> context,
+    int32_t world_id) {
+  if (page_changed_callback_.IsEmpty()) {
+    return;
+  }
+  // If we have a page changed callback, and its creation context is the context
+  // that is being released, then release the persitant handle to the callback.
+  auto* isolate = GetIsolate(render_frame());
+  v8::Isolate::Scope isolate_scope(isolate);
+  v8::HandleScope handle_scope(isolate);
+  auto function = page_changed_callback_.Get(isolate);
+  if (function->GetCreationContextChecked(isolate) == context) {
+    page_changed_callback_.Reset();
+  }
+}
+
+void CreatorDetectionAgent::OnDestruct() {
+  delete this;
+}
+
+mojo::AssociatedRemote<mojom::CreatorDetectionHost>&
+CreatorDetectionAgent::GetDetectionHost() {
+  if (!detection_host_) {
+    render_frame()->GetRemoteAssociatedInterfaces()->GetInterface(
+        &detection_host_);
+    detection_host_.reset_on_disconnect();
+  }
+
+  return detection_host_;
+}
+
+void CreatorDetectionAgent::MaybeInjectDetectionScript() {
+  if (!page_changed_callback_.IsEmpty()) {
+    return;
+  }
+
+  auto script = GetDetectionScript(render_frame());
+  if (!script) {
+    return;
+  }
+
+  auto* isolate = GetIsolate(render_frame());
+  v8::Isolate::Scope isolate_scope(isolate);
+  v8::HandleScope handle_scope(isolate);
+
+  v8::Local<v8::Context> context;
+  if (script->should_run_in_isolated_world) {
+    context = render_frame()->GetWebFrame()->GetScriptContextFromWorldId(
+        isolate, isolated_world_id_);
```

Comment:
I tested with shields down, and the isolated world seemed to work. Is there some other scenario where the isolated world would not be created when `DidCommitProvisionalLoad` is executed?

Reply:
The world is actually initialized here:
https://github.com/brave/brave-core/blob/master/components/cosmetic_filters/renderer/cosmetic_filters_js_render_frame_observer.cc#L39-L48

Many other features just use it.  Maybe it's okay to just add a comment that it's used by other feature, so it should be no-opt.

Reply:
isolated worlds are normally created when you execute javascript for them. If you need to make sure the isolated world is created then you should explicitly initialize it like CF does above. I don't think it's safe to just assume it will be created somewhere else.